### PR TITLE
Fix 25648 improve read csv stacktrace on bad argument

### LINKED ISSUE #25648: ERR: read_csv exposes an internal function when bad argument is specified
When a user tries to specify a non-existent argument in `read_csv` the error message pertains to what seems to be an internal function (`parser_f`).  This behavior might be a bit surprising / confusing to users, who would expect the error message to refer to the function actually being used.

```python
import pandas as pd
df = pd.read_csv("some_file.csv", bad_arg=True)
```

yields

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-d4606a60902f> in <module>()
----> 1 df = pd.read_csv("some_file.csv", bad_arg=True)

TypeError: parser_f() got an unexpected keyword argument 'bad_arg'
```

### PR DESCRIPTION
**Pull request by roberthdevries**
- [x] closes #25648
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by roberthdevries** [comment]
It works, but it may not be pretty

**Comment by WillAyd** [comment]
Yea this is pretty hacky :-X is this really the only option?

File: doc/source/whatsnew/v1.1.0.rst [hunk 1]
```diff
@@ -355,6 +355,7 @@ I/O
 - Bug in :meth:`read_csv` was causing a segfault when there were blank lines between the header and data rows (:issue:`28071`)
 - Bug in :meth:`read_csv` was raising a misleading exception on a permissions issue (:issue:`23784`)
 - Bug in :meth:`read_csv` was raising an ``IndexError`` when header=None and 2 extra data columns
+- `TypeError` exceptions raised by :meth:`read_csv` and :meth:`read_table` where showing as ``parser_f`` when an unexpected keyword argument was passed (:issue:`25648`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v1.1.0.rst]:
```suggestion
- `TypeError` exceptions raised by :meth:`read_csv` and :meth:`read_table` were showing as ``parser_f`` when an unexpected keyword argument was passed (:issue:`25648`)
```

**Review by gfyoung — CHANGES_REQUESTED** [review]
-1 on this change unfortunately.  `exec` is extremely hacky for something that is arguably minor.

While I don't have an immediate alternative off the top of my head, I think it's better to let this one percolate a little than to force a fix.

**Comment by roberthdevries** [comment]
The options to fix this are limited by the fact that the exception raising mechanism uses a read-only variable of the code attribute of a function: `<func>.code.co_name`.
This variable is set by the python compiler/interpreter by taking the name of the function when it is defined: `def <func>`. Which means that we **have** to do `def read_csv()` and `def read_table()` to fix this issue.

I see 4 options in total:

1. Use `exec()` to dynamically create the functions with the appropriate names
2. Make two duplicates of `parser_f` with all of its arguments, factor out the body of the function
3. Make a template function e.g. `parser_f.py.in` and generate `read_csv.py` and `read_table.py` at build time
4. Don't fix this issue :smiley: 

May I have your suggestions/votes please

**Comment by gfyoung** [comment]
1. -1: too hacky as already discussed
2. -1: maintenance burden > error correctness
3. -0.5: less maintenance burden, but I still feel creating a template just for this is too much

4. +0: It's worked us for awhile, but I don't want to give up on this just yet.

Middle ground option:

Rename `parser_f` to something like `read_document` ?   Then the function name isn't as foreign sounding and at least aligns with what `read_csv` and `read_table` do.

Another middle ground option:

Create a wrapper around `parser_f` than takes in `*args` and `**kwargs` and catch `TypeError` to replace instances of the string `parser_f` with the function name.

The disadvantage is we lose the signature when we do `help` on these functions.

**Comment by jreback** [comment]
agree with comments here - we cannot use string execution like this 

you an try setting 

parser_f.__name__ = ...

**Comment by gfyoung** [comment]
> parser_f.name = ...

As the OP

**Comment by jbrockmendel** [comment]
@roberthdevries can you rebase

**Comment by gfyoung** [comment]
> @roberthdevries can you rebase

Before rebasing, I think it's only fair that we get some consensus on how to address the underlying problem, given that we largely agree that the proposed fix is not something we would like to merge.

**Comment by roberthdevries** [comment]
I am currently ill, and not able to work on this

**Comment by gfyoung** [comment]
> I am currently ill, and not able to work on this

@roberthdevries : Oh no!  Very sorry to hear that.  Hope you get better soon.

We can take over this PR once we get some consensus.

**Comment by jbrockmendel** [comment]
is it possible compat.set_function_name may work where just setting `func.__name__` doesnt?

**Comment by gfyoung** [comment]
> is it possible compat.set_function_name may work where just setting `func.__name__` doesnt?

@jbrockmendel : Unfortunately not.  As mentioned in the OP, the function name in the error message is populated from a read-only attribute.

**Comment by jbrockmendel** [comment]
OK, then im inclined towards the emerging consensus of "learn to live with it"

**Comment by gfyoung** [comment]
> OK, then im inclined towards the emerging consensus of "learn to live with it"

Learn to live with it?  We've been living with this stacktrace issue for years 😄

**Comment by simonjayhawkins** [comment]
@roberthdevries Thanks for looking into this. closing for now as stale. ping if you want to continue.

**Comment by roberthdevries** [comment]
Can you mark  #25648 as wontfix?

**Comment by simonjayhawkins** [comment]
added a needs discussion label and removed the contributions welcome for now.

**Pull request closed without merging**